coderd/files/cache.go

```diff
@@ -0,0 +1,107 @@
+package files
+
+import (
+	"bytes"
+	"context"
+	"io/fs"
+	"sync"
+	"sync/atomic"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+
+	archivefs "github.com/coder/coder/v2/archive/fs"
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/util/lazy"
+)
+
+// NewFromStore returns a file cache that will fetch files from the provided
+// database.
+func NewFromStore(store database.Store) Cache {
+	fetcher := func(ctx context.Context, fileID uuid.UUID) (fs.FS, error) {
+		file, err := store.GetFileByID(ctx, fileID)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to read file from database: %w", err)
+		}
+
+		content := bytes.NewBuffer(file.Data)
+		return archivefs.FromTarReader(content), nil
+	}
+
+	return Cache{
+		lock:    sync.Mutex{},
+		data:    make(map[uuid.UUID]*cacheEntry),
+		fetcher: fetcher,
+	}
+}
+
+// Cache persists the files for template versions, and is used by dynamic
+// parameters to deduplicate the files in memory. When any number of users opens
+// the workspace creation form for a given template version, it's files are
+// loaded into memory exactly once. We hold those files until there are no
+// longer any open connections, and then we remove the value from the map.
+type Cache struct {
+	lock sync.Mutex
+	data map[uuid.UUID]*cacheEntry
+	fetcher
+}
+
+type cacheEntry struct {
+	refCount *atomic.Int64
+	value    *lazy.ValueWithError[fs.FS]
+}
+
+type fetcher func(context.Context, uuid.UUID) (fs.FS, error)
+
+// Acquire will load the fs.FS for the given file. It guarantees that parallel
+// calls for the same fileID will only result in one fetch, and that parallel
+// calls for distinct fileIDs will fetch in parallel.
+func (c *Cache) Acquire(ctx context.Context, fileID uuid.UUID) (fs.FS, error) {
```

Comment:
Do you think it would be a nice helper to add the `Release` call in the return?

```golang
func (c *Cache) Acquire(ctx context.Context, fileID uuid.UUID) (files fs.FS, release func(), err error)
```

so something like:
```golang
	files, err = c.prepare(ctx, fileID).Load()
	if err != nil {
		return nil, nil, err
	}
	return files, func() {
		c.Release(fileID)
	}, err
```

So the caller has the release more accessible?

Reply:
All valid points :thinking: 

It's hard to think about the API right now without it's usage. I am just worried about the caller forgetting to call `Release`.

I'm ok to ship it for now, and we can tweak the api in the future. One idea I had to was to throw it in a closure. Idk if it would work, but it might be possible to throw the websocket and preview all in the closure, and then `Release` will guarantee'd to be called :shrug: 

```golang
	c.Acquire(id, func(fs fs.FS) {
		// Use the FS here, when the closure ends, the FS is `Released`
	})
```